docs: add chainlink-integration guide (#388)
Agoric/documentation: main/chainlink-integration.md

# Chainlink Integration

> This document explains how you can consume [Chainlink

Reviewer: ```suggestion
This document explains how to consume [Chainlink
```

"you can consume" reads like Chainlink oracles are a crunchy snack you can chow down on. : -)

oracles](https://chain.link) when unit testing, integration testing, and
actually on-chain.

## Overview

Using Chainlink on Agoric provides two main features:

> - price feeds exposed on Agoric via the on-chain `home.priceAuthority`, an
>   officially-sponsored [price authority](/zoe/guide/price-authority.md) built
>   from aggregating several Chainlink nodes.

Reviewer: ```suggestion
- Price feeds are exposed on Agoric via the on-chain `home.priceAuthority`. This is an
  officially-sponsored [price authority](/zoe/guide/price-authority.md) built
  from aggregating several Chainlink nodes.
```

- [Chainlink's Any API](https://docs.chain.link/docs/request-and-receive-data)

>   to initiate a job on a single oracle and return its results

Reviewer: This doesn't sync with the intro to it, which is "Using Chainlink provides two main features:".  Maybe something like change "to initiate" to "can initiiate"

> These features have been tested with [actual Chainlink oracle

Reviewer: ```suggestion
We have tested these features with [actual Chainlink oracle
```

software](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).

> However, it is important to note that Chainlink is still in the process of
> setting up an incentivized testnet for established Chainlink node operators to
> connect to Agoric.

Reviewer: ```suggestion
**Note**: Chainlink has not yet finished setting up an incentivized testnet for established Chainlink node operators to connect to Agoric.
```

Maybe add an ETA?

## Price Authority

To test your contract against a locally-simulated price authority, just follow
the instructions in [the Price Authority API](/zoe/guide/price-authority.md).

> If you want to use the curated on-chain price authority, you can find it at
> `home.priceAuthority`.  Here is a sample of getting a quote for selling `30
> Testnet.$LINK` in `Testnet.$USD`:

Reviewer: ```suggestion
To use the curated on-chain price authority, see `home.priceAuthority`.  For example, to get a quote for selling `30
Testnet.$LINK` in `Testnet.$USD`:
```

(Note that this is a mock price until there are actual Chainlink nodes on the
testnet).

```js
const linkIssuer = E(home.wallet).getIssuer('Testnet.$LINK');
const linkMath = await makeLocalAmountMath(linkIssuer);
const linkBrand = await E(linkIssuer).getBrand();
const { decimalPlaces: linkDecimals } = await E(linkBrand).getDisplayInfo();
const linkAmount = linkMath.make(30 * 10 ** linkDecimals);
const usdBrand = await E(E(home.wallet).getIssuer('Testnet.$USD')).getBrand();
const { decimalPlaces: usdDecimals } = await E(usdBrand).getDisplayInfo();
const { quoteAmount: { value: [{ amountOut: usdAmount, timestamp }] } } = await E(home.priceAuthority).quoteGiven(linkAmount, usdBrand);
const linkValue = linkAmount.value / 10 ** linkDecimals;
const usdValue = usdAmount.value / 10 ** usdDecimals;
console.log(linkValue, 'Testnet.$LINK trades for', usdValue, 'Testnet.$USD');
// 30 Testnet.$LINK trades for 12 Testnet.$USD
```

## Any API

> In order to use Chainlink's Any API, you will need to obtain an instance of the

Reviewer: ```suggestion
To use Chainlink's Any API, you need to get an instance of the
```

[Low-level Oracle Query Contract](/zoe/guide/contracts/oracle.md) and submit a
query of the form:

```js
{
  jobId: <Chainlink JobId>,
  params: { ...<job parameters> }
}
```

> The string `result` data that the oracle node returns will be sent as the
> response to your query.

Reviewer: ```suggestion
The string `result` data the oracle node returns is sent as the
response to your query.
```

(Note that this is a mock price until there are actual Chainlink nodes on the
testnet).

```js
const linkIssuer = E(home.wallet).getIssuer('Testnet.$LINK');
const linkMath = await makeLocalAmountMath(linkIssuer);
const linkBrand = await E(linkIssuer).getBrand();
const { decimalPlaces: linkDecimals } = await E(linkBrand).getDisplayInfo();
const linkAmount = linkMath.make(30 * 10 ** linkDecimals);
const usdBrand = await E(E(home.wallet).getIssuer('Testnet.$USD')).getBrand();
const { decimalPlaces: usdDecimals } = await E(usdBrand).getDisplayInfo();
const { quoteAmount: { value: [{ amountOut: usdAmount, timestamp }] } } = await E(home.priceAuthority).quoteGiven(linkAmount, usdBrand);
const linkValue = linkAmount.value / 10 ** linkDecimals;
const usdValue = usdAmount.value / 10 ** usdDecimals;
console.log(linkValue, 'Testnet.$LINK trades for', usdValue, 'Testnet.$USD');
// 30 Testnet.$LINK trades for 12 Testnet.$USD
```

## Any API

To use Chainlink's Any API, you need to get an instance of the
[Low-level Oracle Query Contract](/zoe/guide/contracts/oracle.md) and submit a
query of the form:

```js
{
  jobId: <Chainlink JobId>,
  params: { ...<job parameters> }
}
```

> The string `result` data the oracle node returns is sent as the

Reviewer: ```suggestion
The string `result` data the oracle node returns is sent as the
```
I'm not quite following what you mean by "The string 'result' data". Is the returned data just 'result'? Is it data prefaced by 'result'? In general, I'd like to change the sentence, depending on what those first four words mean, to something like:
"The oracle node returns <whatever>. This is sent as your query's response."

Author: Thanks, I've clarified it to: "The oracle node returns its result, which is conventionally a string. This is sent as your query's reply."

response to your query.

> You can test these queries against a locally-running Chainlink node that you
> control by following [the Chainlink integration
> instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).

Reviewer: ```suggestion
You can test these queries against a locally-running Chainlink node. You
control the node by following [the Chainlink integration
instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md).
```

Author: ```suggestion
You can test these queries against a locally-running Chainlink node that you
control.  Follow [the Chainlink integration
instructions](https://github.com/Agoric/dapp-oracle/blob/main/chainlink-agoric/README.md) to set it up.
```

> If you want to use a more limited local node that emulates just a portion of the
> Chainlink API without having to run Docker containers, you can use the [local
> (mock) builtin
> oracle](https://github.com/agoric/dapp-oracle#running-a-local-builtin-oracle).

Reviewer: ```suggestion
There is a more limited local node that emulates part of the
Chainlink API without having to run Docker containers.  This is the [local
(mock) builtin
oracle](https://github.com/agoric/dapp-oracle#running-a-local-builtin-oracle).
```
